pkg/controller/anomalydetector/controller.go

```diff
@@ -191,6 +204,134 @@ func (c *AnomalyDetectorController) Run(stopCh <-chan struct{}) {
 	<-stopCh
 }
 
+func (c *AnomalyDetectorController) gcworker(stopCh <-chan struct{}) {
+	wait.PollImmediateUntil(time.Second, func() (done bool, err error) {
+		return !c.processNextGcWorkItem(), nil
+	}, stopCh)
+}
+
+func (c *AnomalyDetectorController) processNextGcWorkItem() bool {
+	obj, quit := c.gcQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gcQueue.Done(obj)
+
+	if key, ok := obj.(controllerutil.GcKey); !ok {
+		c.queue.Forget(obj)
+		klog.ErrorS(nil, "Expected gcKey in work queue", "got", obj)
+		return false
+	} else if updatedKey, err := c.handleStaleResources(key); err == nil {
+		c.gcQueue.Forget(key)
+		return false
+	} else {
+		klog.ErrorS(err, "Error handling stale resources, requeuing it")
+		c.gcQueue.AddRateLimited(updatedKey)
+	}
+	return true
+}
+
+// handleStaleResources handles the stale Spark Applications and database entries.
+func (c *AnomalyDetectorController) handleStaleResources(key controllerutil.GcKey) (updatedKey controllerutil.GcKey, err error) {
```

Comment:
Is it possible to reuse the series of resync functions (handleStaleResources, handleStaleSparkApp, handleStaleSparkApp) for both the anomaly detector and policy recommendation? It would be more efficient if we could avoid duplicating this code every time a new Spark application is involved.
cc. @yanjunz97

Reply:
I think we can use interface for when we have to call a common function and not from within the controller, however that might not work here
e.g. if we have two controllers controller1 and controller 2, we can use a same function for both and pass the struct variables and it should work perfectly but we are calling controller function (GetThroughputAnomalyDetector/GetNetworkPolicyRecommendation) within the common function which has a different return type than the other, which I guess won’t be feasible to pass function as an argument to a common function as diff return type would cause issue, so I guess best practice would be to duplicate the code.
Please correct me if I am wrong.

Reply:
As the get and list functions for different resources still have the same number of parameter, maybe you can pass `function(xxx) interface{}` and a string (or type) to indicate the resource as arguments and cast the return value according to the resource in the common function.